### ISSUE TITLE
Destroying entities might add them to groups

### ISSUE THREAD

**Issue by sschmid** [sschmid/Entitas-CSharp#192]
Using a NoneOf matcher for groups can lead to adding entities to a group in the process of destroying it.
E.g.
```
Matcher.AllOf(Position).NoneOf(Destroy)
```

When destroying an entity all components will be removed. Depending on the order of removing those components it can happen that the entity will be added to the group. If DestroyComponent gets removed first the entity will matches the matcher. This shouldn't happen.

**Comment by alandyshev** [comment]
PrimaryEntityIndex throws exception due to this bug:
```
EntityIndexException: Entity for key '0' already exists!
Only one entity for a primary key is allowed.
Entitas.PrimaryEntityIndex`1[System.Int32].addEntity (Entitas.Entity entity, IComponent component) (at Assets/Entitas/EntityIndex.cs:108)
Entitas.AbstractEntityIndex`1[System.Int32].onEntityAdded (Entitas.Group group, Entitas.Entity entity, Int32 index, IComponent component) (at Assets/Entitas/EntityIndex.cs:40)
Entitas.Pool.updateGroupsComponentAddedOrRemoved (Entitas.Entity entity, Int32 index, IComponent component) (at Assets/Entitas/Pool.cs:319)
Entitas.Entity.replaceComponent (Int32 index, IComponent replacement) (at Assets/Entitas/Entity.cs:155)
Entitas.Entity.RemoveAllComponents () (at Assets/Entitas/Entity.cs:253)
Entitas.Entity.destroy () (at Assets/Entitas/Entity.cs:353)
Entitas.Pool.DestroyEntity (Entitas.Entity entity) (at Assets/Entitas/Pool.cs:146)
DestroyEntitySystem.Execute (System.Collections.Generic.List`1 entities) (at Assets/Sources/Common/Core/Systems/DestroyEntitySystem.cs:21)
Entitas.ReactiveSystem.Execute () (at Assets/Entitas/ReactiveSystem.cs:118)
```

Here's my code for EntityIndex:
```
    // Tube Segment
    public const string TubeSegmentWithIdKey = "TubeSegmentWithId";
    private static void AddTubeSegmentWithIdEntityIndex(Pools pools)
    {
        var index = new PrimaryEntityIndex<int>(
            pools.pool.GetGroup(Matcher.AllOf(Matcher.TubeSegment, Matcher.Id).NoneOf(Matcher.Destroy)),
            (entity, component) =>
            {
                var idComponent = component as IdComponent;
                return idComponent != null
                    ? idComponent.value
                    : entity.id.value;
            });
        pools.pool.AddEntityIndex(TubeSegmentWithIdKey, index);
    }

    public static Entity GetTubeSegmentWithId(this Pool pool, int id) {
        var index = (PrimaryEntityIndex<int>)pool.GetEntityIndex(TubeSegmentWithIdKey);
        return index.HasEntity(id) ? index.GetEntity(id) : null;
    }
```

**Comment by sschmid** [comment]
@alandyshev let me know if there are still issues

**Comment by sschmid** [comment]
Released [Entitas 0.35.0](https://github.com/sschmid/Entitas-CSharp/releases/tag/0.35.0) with the fix.

**Comment by alandyshev** [comment]
It's working now. Exception is not thrown. Thanks.
